packages/trie/src/trie.ts

```diff
@@ -278,68 +278,63 @@ export class Trie {
-              resolve({ node: null, remaining: keyRemainder, stack })
-            } else {
-              // node found, continuing search
-              // this can be optimized as this calls getBranch again.
-              walkController.onlyBranchIndex(node, keyProgress, branchIndex)
-            }
-          }
-        } else if (node instanceof LeafNode) {
-          if (doKeysMatch(keyRemainder, node.key())) {
-            // keys match, return node with empty key
-            resolve({ node, remaining: [], stack })
-          } else {
-            // reached leaf but keys dont match
-            resolve({ node: null, remaining: keyRemainder, stack })
-          }
-        } else if (node instanceof ExtensionNode) {
-          const matchingLen = matchingNibbleLength(keyRemainder, node.key())
-          if (matchingLen !== node.key().length) {
-            // keys don't match, fail
-            resolve({ node: null, remaining: keyRemainder, stack })
+    const onFound: FoundNodeFunction = async (_, node, keyProgress, walkController) => {
+      // If we already have a result, exit early
+      if (result) return
+
+      if (node === null) {
+        result = { node: null, remaining: [], stack }
+        return
+      }
+
+      const keyRemainder = targetKey.slice(matchingNibbleLength(keyProgress, targetKey))
+      stack.push(node)
+
+      if (node instanceof BranchNode) {
+        if (keyRemainder.length === 0) {
+          result = { node, remaining: [], stack }
+        } else {
+          const branchIndex = keyRemainder[0]
+          const branchNode = node.getBranch(branchIndex)
+          if (!branchNode) {
+            result = { node: null, remaining: keyRemainder, stack }
           } else {
-            // keys match, continue search
-            walkController.allChildren(node, keyProgress)
+            walkController.onlyBranchIndex(node, keyProgress, branchIndex)
           }
         }
-      }
-
-      // walk trie and process nodes
-      try {
-        await this.walkTrie(this.root(), onFound)
-      } catch (error: any) {
-        if (error.message === 'Missing node in DB' && !throwIfMissing) {
-          // pass
+      } else if (node instanceof LeafNode) {
+        if (doKeysMatch(keyRemainder, node.key())) {
+          result = { node, remaining: [], stack }
         } else {
-          reject(error)
+          result = { node: null, remaining: keyRemainder, stack }
+        }
+      } else if (node instanceof ExtensionNode) {
+        const matchingLen = matchingNibbleLength(keyRemainder, node.key())
+        if (matchingLen !== node.key().length) {
+          result = { node: null, remaining: keyRemainder, stack }
+        } else {
+          walkController.allChildren(node, keyProgress)
         }
       }
+    }
 
-      // Resolve if walkTrie finishes without finding any nodes
-      resolve({ node: null, remaining: [], stack })
-    })
+    try {
+      await this.walkTrie(this.root(), onFound)
```

Comment:
Is the further goal of "simplification" to have `walkTrie` return the actual result instead of binding `onFound` to the outer JS closure?

Reply:
I would call that *a* further goal...